### DISCUSSION TITLE
Custom agent (like MemGPT) support in AutoGen Studio

### DISCUSSION THREAD

**Discussion by lpezet** [microsoft/autogen#1438] [Ideas]
Hello!

I would like to know if there's any plan to support some more dynamic/custom agents in Autogen Studio (with MemGPT in mind here)?
Looking at [workflowmanager.py](https://github.com/microsoft/autogen/blob/main/samples/apps/autogen-studio/autogenstudio/workflowmanager.py#L174):
```
        if agent_type == "assistant":
            agent = autogen.AssistantAgent(**agent_config.dict())
            agent.register_reply([autogen.Agent, None], reply_func=self.process_reply, config={"callback": None})
        elif agent_type == "userproxy":
            agent = autogen.UserProxyAgent(**agent_config.dict())
            agent.register_reply([autogen.Agent, None], reply_func=self.process_reply, config={"callback": None})
        else:
            raise ValueError(f"Unknown agent type: {agent_type}")
        return agent
```
I'm just wondering 1) whether or not you guys planned on anything there and 2) would you go with some kind of dynamic loading/plugin-style approach here to load/construct an agent?
For 2), I mean something like this (this is crude, just to demonstrate):
```
        if agent_type == "assistant":
            agent = autogen.AssistantAgent(**agent_config.dict())
            agent.register_reply([autogen.Agent, None], reply_func=self.process_reply, config={"callback": None})
        elif agent_type == "userproxy":
            agent = autogen.UserProxyAgent(**agent_config.dict())
            agent.register_reply([autogen.Agent, None], reply_func=self.process_reply, config={"callback": None})
	# == BEGIN CHANGES ==
	elif module_exists(agent_type):
		module = importlib.import_module(...)
		AgentClass = getattr(module, ...)
		agent = AgentClass(**agent_config.dict())
		agent.register_reply([autogen.Agent, None], reply_func=self.process_reply, config={"callback": None})
	# == END CHANGES ==
        else:
            raise ValueError(f"Unknown agent type: {agent_type}")
        return agent
```

**Comment by perrosnk** [comment]
I am also interested in this.

**Comment by sonichi** [comment]
@victordibia @gagb @ekzhu

**Comment by victordibia** [comment]
@ipezet, 

This is a great idea.
We definitely like to support custom agents in AutoGen studio. 
The AutoGen studio roadmap is here - #737  and support for other agents is on the list.
Currently, the DSL and abstractions for two agent and groupchat patterns are still getting stabilized. Once this works reasonably well, we'll focus more on additional agent classes.

## Loading Arbitrary Agent Classes (Based on ConversableAgent) is a Good Idea
I really like the direction of your general proposal on extending workflowmanager to load arbitrary agent classes. 
In general such an agent class will fit in well with the rest of autogenstudio if it inherits from the ConversableAgent class. 

## PR Welcome!

If you have a use case with a potential implementation above, I'll be happy to discuss, improve and merge.

Please note that development for autogenstudio is done on the autogenstudio branch (which is usually significantly ahead of main). Please base your changes on that branch, following the [autogenstudio contribution guide](https://github.com/microsoft/autogen/tree/autogenstudio/samples/apps/autogen-studio#contribution-guide).
